kraken-build/src/kraken/std/cargo/manifest.py

```diff
@@ -79,6 +79,15 @@ def read(cls, project_dir: Path) -> CargoMetadata:
             "--manifest-path",
             str(project_dir / "Cargo.toml"),
         ]
+        if locked is None:
+            for parent in (project_dir / "Cargo.toml").absolute().parents:
```

Comment:
Yep, I buy those arguments!

One last question: The join with `Cargo.toml` is only to make sure we also check `project_dir`, yes? In that case, I think this is less distracting:

```suggestion
            for parent in (project_dir, *project_dir.parents):
```

Reply:
Yep, that's why, although I think we still need `.absolute()`, because otherwise `.parents` would be empty when `project_dir` is `.` for example. At least my understand of `parents` is that it operates on `dir.split("/")` rather than resolving `..` at each step (but I could be wrong about that).

Also, wouldn't `()` need to become `[]`?

Reply:
Agreed, we should keep it. I accidentally dropped it from the example!

Reply:
Is
```rust
for parent in [project_dir, *project_dir.absolute().parents]:
```
clearer than what's there now though?

Reply:
I think the intent is clearer, yes. Could move the `.absolute()` to a separate line to make it more readable as well

Reply:
```suggestion
            project_dir = project_dir.absolute()
            for parent in [project_dir, *project_dir.parents]:
```